core/sr-api-macros/tests/runtime_calls.rs

```diff
@@ -191,3 +192,38 @@ fn record_proof_works() {
 		&block.encode(),
 	).expect("Executes block while using the proof backend");
 }
+
+#[test]
+fn returns_mutable_static() {
+	let client = TestClientBuilder::new().set_execution_strategy(ExecutionStrategy::AlwaysWasm).build();
+	let runtime_api = client.runtime_api();
+	let block_id = BlockId::Number(client.info().chain.best_number);
+
+	let ret = runtime_api.returns_mutable_static(&block_id).unwrap();
+	assert_eq!(ret, 33);
+
+	// We expect that every invocation will need to return the initial
+	// value plus one. If the value increases more than that then it is
+	// a sign that the wasm runtime preserves the memory content.
+	let ret = runtime_api.returns_mutable_static(&block_id).unwrap();
+	assert_eq!(ret, 33);
+}
+
+// If we didn't restore the wasm instance properly, on a trap the stack pointer would not be
+// returned to its initial value and thus the stack space is going to be leaked.
+//
+// See https://github.com/paritytech/substrate/issues/2967 for details
+#[test]
+fn restoration_of_globals() {
+	let client = TestClientBuilder::new().set_execution_strategy(ExecutionStrategy::AlwaysWasm).build();
+	let runtime_api = client.runtime_api();
+	let block_id = BlockId::Number(client.info().chain.best_number);
+
+	// On the first invocation we allocate approx. 75% of stack and then trap.
+	let ret = runtime_api.allocates_stack_alot(&block_id, true);
+	assert!(ret.is_err());
+
+	// On the second invocation we allocate yet another 75% of stack 
+	let ret = runtime_api.allocates_stack_alot(&block_id, false);
+	assert!(ret.is_ok());
+}
```

Comment:
This place is super convenient to place these tests. However, these tests test the runtime behavior and it feels weird to place them here. Do you have a better idea where to put them @bkchr ?

Reply:
Yeah, that is correct. I used this here to test the actual behavior of `sr-api-macros`. 

Why not put the tests directly into `test-runtime`? I think that is the best place.

Reply:
But can we do that? These test use `TestClient` which is from `test-client` crate which depends on the `test-runtime`...

Reply:
Yes, when you put this into `dev-dependencies`. We do this in multiple places in Substrate^^